parquet-column/src/main/java/org/apache/parquet/column/values/factory/DefaultV1ValuesWriterFactory.java

```diff
@@ -100,11 +101,17 @@ private ValuesWriter getInt96ValuesWriter(ColumnDescriptor path) {
   }
 
   private ValuesWriter getDoubleValuesWriter(ColumnDescriptor path) {
+    if (this.parquetProperties.isByteStreamSplitEnabled()) {
```

Comment:
I changed it so that ByteStreamSplit is only used if dictionary is not enabled and the bytestreamsplit is enabled.
The BYTE_STREAM_SPLIT encoding only makes sense when the floating-point data has high enough entropy and also it has to combined with a compressor.

Reply:
Dictionary encoding works the following way. Dictionary encodes the incoming values until all the values are written or if the dictionary gets full (there is a size limit) then the actual page is re-encoded to the fallback encoder and then all the left values are encoded with it. It means that some (or 0) pages are dictionary encoded and some (or 0) are encoded with the fallback encoder.
So, I think BYTE_STREAM_SPLIT should be used as the fallback encoder instead of the plain encoder in case of the related property is true.

Reply:
I think it would be better if byte_stream_split is used when dictionary is not enabled and byte_stream_split is requested, not as a fallback to dictionary.
In my evaluation byte_stream_split made sense for cases when the data has high enough entropy: ~15bits and higher. This included almost all samples of scientific data from our gathered data set.
I also did an artificial test to compare the Dictionary encoding with this one as the entropy of the data increases: https://github.com/martinradev/arrow-fp-compression-bench/blob/master/entropy_ratio.jpeg
It is noticeable that the Dictionary encoder falls back to the plain encoder around 17 bits of entropy but that the byte_stream_split encoding is superior for data with less entropy.
Also, the byte_stream_split encoding/decoding is essentially a scatter/gather operation, so it's as fast as memcpy. The dictionary encoding implementation is not as fast as memcpy:
https://github.com/martinradev/arrow-fp-compression-bench/blob/master/entropy_write_speed.jpeg
Note that the latter plot is not my optimized simd implementation.
I do not know how the results would look for the java implementation.

So, I think that the control should be left in the user's hands and I think it makes sense to sanitize building a parquet writer to disallow requesting both dictionary and the byte_stream_split encoding.

Reply:
I have some problems with this approach:
* parquet-mr tries to figure out the correct encoding. It might not be able to always choose the best but the idea of dictionary+fallback follows this approach. If the dictionary encoding is not good enough (because of the high entropy) it will fallback to another encoding. The fallback might happen even at the first page meaning that no dictionary encoding takes place for that column. It seems that byte_stream_split+compression is always a better choice than plain+compression so why would we use it?
* The fallback happens because of the dictionary page gets full. The size of the dictionary page is configurable so the entropy when the fallback happens is not a fix border and is different by default for float and double.
* Currently, you are not able to set dictionary encoding (or anything else) separately for the different columns. So, following your approach you have to decide if you want byte_stream_split encoding for your fp column or you want dictionary encoding for your string column but you cannot have both.

I would't say I know the best approach here. It might be a good idea to propose a new framework and configuration to allow the user to set the exact encoding/compression etc. for each column but it is a much bigger issue.
If we agree on that byte_stream_split is always better than plain, let's simply drop using plain encoding and use byte_stream_split in its place. If you would like to allow the user to enforce byte_stream_split over dictionary in all cases then you can have a property for that but what if you have more than one fp column with different entropy?